Sprint _10

Wilwous opened the pull request:
Работа для сдачи 10 спринта

Wilwous committed:
UPD: Сверстал с сториборде AuthViewController

Wilwous committed:
Поправил констрейны кнопки войти

Wilwous committed:
UPD2: Сверстал с сториборде AuthViewController

Wilwous committed:
Сверстал часть WebViewViewController

Wilwous committed:
Сверстал WebViewViewController

Wilwous committed:
Сверстал WebViewViewController

Wilwous committed:
Добавил кнопку назад в WebViewViewController

Wilwous committed:
Добавил URLComponents и добавил скругление кнопки войти

Wilwous committed:
Исправил ошибку

Wilwous committed:
Добавил WKNavigationDelegate

Wilwous committed:
Поправил код, вынес протокол WebViewViewControllerDelegate в отдельный файл и добавил делегацию AuthViewController в WebViewViewController.

Wilwous committed:
Поправил код, вынес протокол WebViewViewControllerDelegate в отдельный файл и добавил делегацию AuthViewController в WebViewViewController.

Wilwous committed:
Добавил UIProgressView

Wilwous committed:
Поправил констрейны

Wilwous committed:
UPD: Поправил констрейны

Wilwous committed:
Улучшена интеграция индикатора прогресса в методы viewWillAppear и viewWillDisappear, а также в метод observeValue для отслеживания изменений свойства estimatedProgress объекта WKWebView. Добавлена обработка нажатия на кнопку "Назад" с вызовом метода делегата webViewViewControllerDidCancel.

Wilwous committed:
Вынес структуру URLComponents в отдельную функцию в расширение класса

Wilwous committed:
Добавил во все файлы Mark и организовал код

Wilwous committed:
Добавлен OAuth2Service для запроса токена доступа

Wilwous committed:
Добавлена OAuth2 авторизация

Wilwous committed:
Создан SplashViewController

Wilwous committed:
UPD: Создан SplashViewController

Wilwous committed:
Назначил класс в контроллере

helecta commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,50 @@
+    private var showWebViewSegueIdentifier = "ShowWebView"
+    private let oauth2Service = OAuth2Service()
+    
+    weak var delegate: AuthViewControllerDelegate?
```
Стоит организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

helecta commented on ImageFeed/Auth/ServiceAccessLayer/OAuth2Service.swift:
```diff
@@ -0,0 +1,142 @@
+import Foundation
+
+private struct OAuthTokenResponseBody: Decodable {
+    let access_token: String
```
В swift принято использовать camelCase, а не snake_case
Советую почитать https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable  и исправить

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,138 @@
+    // MARK: Private Functions
+    
+    private func updateProgress() {
+        progressView.progress = Float(webView.estimatedProgress)
```
если использовать progressView.setProgress(…), то можно сделать заполнение прогресс бара более плавным.

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,138 @@
+    
+    // MARK: Constants
+    
+    let UnsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Такие константы стоит называть со строчной буквы.  Да, в учебнике было так, но все-таки правильно называть со строчной. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on ImageFeed/SplashView/SplashViewController.swift:
```diff
@@ -0,0 +1,78 @@
+
+final class SplashViewController: UIViewController {
+    
+    private let ShowAuthenticationScreenSegueIdentifier = "ShowAuthenticationScreen"
```
Такие константы стоит называть со строчной буквы.  Да, в учебнике было так, но все-таки правильно называть со строчной. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -1,47 +1,59 @@
 import UIKit
 
-final class ImagesListViewController: UIViewController {
+class ImagesListViewController: UIViewController {
```
Почему удалил final?

Ключевое слово final в Swift позволяет объявить класс, свойство или метод, который не может быть переопределён. Это не только защищает твой код от случайного переопределения,
но и помогает компилятору Swift производить дополнительные оптимизации кода,
что улучшает производительность.
Да и просто является хорошим тоном: при ревью сразу считывается намерение разработчика наследоваться от этого класса или нет.
Поэтому тут final очень даже к месту)

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -1,47 +1,59 @@
-            let indexPatch = sender as! IndexPath
-            let image = UIImage(named: photosName[indexPatch.row])
-            viewController.image = image
+            let indexPath = sender as! IndexPath
```
Желательно не использовать force операции, лучше их избегать. Т.к они могут приводить к падению приложения, считается что падение приложения хуже, чем не рабочее состояние.
Тут можно переписать с использование guard например

helecta commented on ImageFeed/Constants.swift:
```diff
@@ -7,8 +7,10 @@
-    let AccessScope = "public+read_user+write_likes"
-    let DefaultBaseURL = URL(string: "https://api.unsplash.com")!
+enum Constants {
+    static let accessKey = "<ItpF13R0bOeRoXFlT7lruRbTWWiXzJmUyMqSy4Q0PZQ>"
```
константы указываются без <>, те:
'    static let accessKey = "<ItpF13R0bOeRoXFlT7lruRbTWWiXzJmUyMqSy4Q0PZQ"'
Аналогично для `secretKey ` и `redirectURI `
Но даже если их указать, то у тебя они все равно не верно указаны. Проверь пожалуйста их еще раз, что ты правильно скопировал свои ключи с сайта unsplash.

Wilwous replied:
Поправил, но ключи я проверил и все указано верно 🤔
Не поставил галочку write likes access.

helecta commented on ImageFeed/SplashView/SplashViewController.swift:
```diff
@@ -0,0 +1,78 @@
+            guard let self = self else { return }
+            switch result {
+            case .success:
+                self.switchToTabBarController()
```
Здесь у тебя приложение упадет, т.к ты вызываешь не в главном потоке

Wilwous replied:
Проверил и блок completion вызван в главном потоке DispatchQueue.main, это можно увидеть в функции data в файле OAuth2Service.

helecta replied:
Да, действительно. Странно я там отловила падение, но может оно было связанно с другими замечаниями, которые ты поправил

Wilwous committed:
Исправление ошибок после ревью:
Поправил констрейны кнопки "Назад"
Убрал ширину кнопки "Войти"
Добавил final в class ImagesListViewController
Поправил константы

Wilwous committed:
Переписал функцию updateProgress с использованием progressView.setProgress для более плавной анимации.

Wilwous committed:
Переписал функцию prepare с использованием безопасного приведения типа.

Wilwous committed:
Переорганизовал код

Wilwous committed:
Добавил слабые ссылки

Wilwous committed:
Пофиксил проблему с авторизацией

Wilwous commented:
Пофиксил все критические ошибки, так же учел рекомендации тут и на платформе (оставил комментарии).
По поводу 4 пункта, я проверил и у меня находиться в главном потоке (приложил скрины).
![photo_2023-10-03 10 05 21](https://github.com/Wilwous/ImageFeed/assets/134493446/f057d5f8-05dc-4584-a254-f66c9565faea)
![photo_2023-10-03 10 05 24](https://github.com/Wilwous/ImageFeed/assets/134493446/30e8f35c-ffa7-4e49-90a8-6234785fa330)

Wilwous committed:
Поправил констрейн

Wilwous committed:
Тест

Wilwous committed:
Тест 2
